heirloom/nohup/nohup.1
.\"
.\" Sccsid @(#)nohup.1	1.5 (gritter) 10/11/03
.\" Parts taken from nice(1), Unix 7th edition:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"   Redistributions of source code and documentation must retain the
.\"    above copyright notice, this list of conditions and the following
.\"    disclaimer.
.\"   Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"   All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed or owned by Caldera
.\"      International, Inc.
.\"   Neither the name of Caldera International, Inc. nor the names of
.\"    other contributors may be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
.\" INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE
.\" LIABLE FOR ANY DIRECT, INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\" BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
.\" OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
.\" EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.TH NOHUP 1 "10/11/03" "Heirloom Toolchest" "User Commands"
.SH NAME
nohup \- run a command immune to hangups
.SH SYNOPSIS
\fBnohup\fI utility \fR[\fIarguments\fR\ ...]
.SH DESCRIPTION
The
.B nohup
command executes
.I utility
immune to hangup signals,
and, for
.B /usr/5bin/nohup
and
.BR /usr/5bin/s42/nohup ,
also immune to quit signals.
If standard output is not re-directed by the user to a non-tty file,
it will be appended to the file `\fInohup.out\fR'.
If \fInohup.out\fR is not writable
in the current directory,
output is appended to
.IR $HOME/nohup.out .
If these files do not exist,
they are created with mode 666 for
.B /usr/5bin/nohup
and
.B /usr/5bin/s42/nohup
or with mode 600 for
.BR /usr/5bin/posix/nohup .
If standard error is not re-directed by the user to a non-tty file,
it will be sent to the same file as standard output.
.SH "ENVIRONMENT VARIABLES"
.TP
.B PATH
Used to locate
.IR utility .
.SH "SEE ALSO"
chmod(1),
nice(1),
sh(1),
signal(2),
execvp(3),
isatty(3)


ex-vi/Changes
Release ...
* When a multicolumn character was replaced by another multicolumn character
  in insert mode, the display was not updated appropriately with terminals
  other than xterm.
* Various rendering errors happened with multicolumn characters if they
  started at an even column (counting from 1 upwards).
* When a multicolumn character was inserted and then replaced, the visual
  screen representation was sometimes not updated accordingly.
* Undoing the replacement of a multicolumn character by a singlecolumn
  character twice made the singlecolumn character invisible.
* The 'cw' visual command with a multibyte character as last character of
  the affected word located at the end of the line left garbage bytes past
  the end of the new text.
* Visual 'U' followed by 'u' lead to garbage on the screen when multibyte
  characters were present on the changed line.
* The handling of multicolumn characters was fixed for terminals without the
  IM (insert mode) capability. It is unlikely that such terminals actually
  exist, but vi will use the corresponding code for open mode if a termcap
  entry is not available.
* When an illegal multibyte sequence is entered in vi insert mode, no garbage
  bytes are generated anymore when the insert mode is left.
* The '\u', '\l', '\U', and '\L' substitution sequences work with multibyte
  characters now.
* Handle character case conversions with the '~' vi command correctly if the
  length of the converted multibyte sequence is smaller than the original one.
* Multibyte sequences that correspond to an unprintable character are now
  printed as multiple octal escape sequences.
* Fixed an old vi bug: If a vi command that yanked or deleted part of a line
  was followed by an ex command that also yanked or deleted some text, a
  following 'p' vi command pasted the text affected by the former vi command.
  It now pastes the text of the last yank or delete even if that was an ex
  command.

Release 12/2/04
* Support for multibyte character locales was added.
* The code has been converted to ANSI C, and support for pre-POSIX systems has
  been dropped.
* When the end of the current line consists of blank characters and the 'w'
  visual command is given at this point, vi now advances to the start of the
  next line instead of ringing the bell. This is compatible with SVR4 vi and
  seems to be what POSIX specifies.
* If the replacement part of a substitute command consists of a single '%',
  as in ':s/foo/%/', the replacement part of the previous substitution is
  used. This is compatible with SVR4 vi and is specified by POSIX.
* Fixed a number of possible heap overflows, e.g. because of too long tag
  strings.

Release 6/5/04
* Some changes were made to the Makefile to support RPM builds. In particular,
  the meaning of the DESTDIR and PREFIX variables was changed.
* An insufficient size of a variable caused the window size set to 8 on
  FreeBSD if the terminal baud rate was 38400.

Release 1/3/04
* Changes to the included libterm only: Made multiple tc= capabilities in
  a termcap entry work; recognize tc= also if it is not the last capability
  in an entry (Bugreport by Andrew Minter).

Release 9/3/03
* The code did not check st_blksize from stat(2) at other points besides
  the one fixed in the last release.
* The keyboard input character with code 255 ("y in ISO-8859-1) was
  misinterpreted as end-of-file indicator.

Release 8/27/03
* Compile fixes for AIX and HP-UX (Mike Jetzer).
* Delete temporary file when preserving was successful after receiving
  SIGHUP or SIGTERM (Fix taken from 4.3BSD Reno).
* Set MAILRC to /dev/null in expreserve to avoid reading the user's
  mail initialization file.
* Optionally use Caldera's 'UNIX(R) Regular Expression Library' to
  get POSIX.2 compatible REs.
* Don't refuse to quit with 'No write since last change' if a line of a
  newly read input file is too long but no changes were made by the user.
* The POSIX_2 preprocessor define has been removed. The behavior previously
  enabled by this variable is now the default (except as noted below).
* Backslash inside RE bracket expresssions (as in [\]]) is now recognized
  as an escape character by default. This is not POSIX.2 compliant, but is
  compatible with historic vi behavior, is demanded by the some traditional
  standards such as the System V Interface definition, and is compatible
  with vim. To get POSIX.2 compliant behavior here, define NO_BE_BACKSLASH.
* The input buffer did overflow with large values of st_blksize from stat(2).

Release 4/3/02
* Option "showmode" has no effect on hardcopy or glass terminals (P).
* Fixed undo for :r! command.
* Interrupt character is not misinterpreted as DEL on insertion (P).
* Fixed interaction of <count>~ vi command with abbreviations and macros (P).
* Avoid horizontal scrolling of showmode string on wraparound to last line (P).
* No showmode string is printed when executing a macro.
* Recovery listing fixed not to mix characters from long filenames in
  /var/preserve with /var/tmp listing (P).
* Catch SIGXFSZ (file size limit exceeded).

Release 3/3/02
* Separated terse/noterse messages for RE errors (P).
* The expreserve and exrecover commands work again (P).
* Passing beyond EOL in <count>~ vi command is not an error (P).
* Fixed segmentation violation on mdjc'd and related bugs (Sven Mascheck).
* Marks remain on lines changed in an undo operation.
* Close mapmalloc file descriptor for /dev/zero on exec() (P).
* Added -L and -V command line options as in SVr4 ex.
* POSIX.2: Processing EXINIT does not depend on stdin being a terminal.
* POSIX.2: No newline is appended to an empty file in visual mode.
* Fixed segmentation violation on :f^V^M^M and similar commands.
* Mapmalloc extended to allocate and release multiple memory pools.
* Exrecover lists /var/tmp in addition to /var/preserve.
* Have multiple attempts to create named buffer temporary file.
* Size limit for temporary file names removed.

Release 2/17/02

heirloom/nohup/nohup.c
/*
 * nohup - run a command immune to hangups
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, August 2002.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute
 * it freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 *
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 *
 * 3. This notice may not be removed or altered from any source distribution.
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
#ifdef	SUS
static const char sccsid[] USED = "@(#)nohup_sus.sl	1.10 (gritter) 4/20/04";
#else
static const char sccsid[] USED = "@(#)nohup.sl	1.10 (gritter) 4/20/04";
#endif

#include	<sys/types.h>
#include	<sys/stat.h>
#include	<fcntl.h>
#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<signal.h>
#include	<libgen.h>
#include	<limits.h>

static char	*progname;		/* argv[0] to main() */
static int	errorfd = 2;

static void
fatal(int eno, const char *string, int val)
{
	char	buf[PATH_MAX];

	snprintf(buf, sizeof buf, "%s: %s: %s\n", progname, string,
				strerror(eno));
	write(errorfd, buf, strlen(buf));
	exit(val);
}

/*
 * Memory allocation with check.
 */
static void *
srealloc(void *vp, size_t nbytes)
{
	void *p;

	if ((p = realloc(vp, nbytes)) == NULL) {
		write(2, "Out of memory\n", 14);
		exit(077);
	}
	return p;
}

static void *
smalloc(size_t nbytes)
{
	return srealloc(NULL, nbytes);
}

static void
usage(void)
{
	fprintf(stderr, "usage: %s command arg ...\n", progname);
	exit(0177);
}

static void
success(const char *fn)
{
	fprintf(stderr, "Sending output to %s\n", fn);
}

static void
failed(const char *fn)
{
	fprintf(stderr, "%s: cannot open/create %s\n", progname, fn);
	exit(0177);
}

static void
create_nohup_out(void)
{
	char	*nohup_out = "nohup.out";
	char	*fn;
	int	fd;
#ifdef	SUS
	const mode_t	mode = S_IRUSR | S_IWUSR;
#else
	const mode_t	mode = 0666;
#endif

	fn = nohup_out;
	if ((fd = open(fn, O_WRONLY|O_CREAT|O_APPEND, mode)) < 0) {
		char	*home;

		if ((home = getenv("HOME")) != NULL) {
			char	*cp, *cq;

			cp = fn = smalloc(strlen(home) + strlen(nohup_out) + 2);
			for (cq = home; *cq; cq++)
				*cp++ = *cq;
			*cp++ = '/';
			cq = nohup_out;
			while ((*cp++ = *cq++) != '\0');
			if ((fd = open(fn, O_WRONLY|O_CREAT|O_APPEND,
							mode)) < 0)
				failed(fn);
		} else
			failed(fn);
	}
	success(fn);
	if (dup2(fd, 1) < 0)
		fatal(errno, "dup2", 0177);
	close(fd);
}

int
main(int argc, char **argv)
{
	int i;

	progname = basename(argv[0]);
	if (argc > 1 && argv[1][0] == '-' && argv[1][1] == '-'
			&& argv[1][2] == '\0')
		argv++, argc--;
	if (argc <= 1)
		usage();
	if (isatty(1))
		create_nohup_out();
	errorfd = dup(2);
	fcntl(errorfd, F_SETFD, FD_CLOEXEC);
	if (isatty(2))
		if (dup2(1, 2) < 0)
			fatal(errno, "dup2", 0177);
	signal(SIGHUP, SIG_IGN);
#ifndef	SUS
	signal(SIGQUIT, SIG_IGN);
#endif
	execvp(argv[1], &argv[1]);
	i = errno;
	fatal(i, argv[1], i == ENOENT ? 0177 : 0176);
	/*NOTREACHED*/
	return 0177;
}
